nvtabular/io/avro.py

```diff
@@ -0,0 +1,138 @@
+        token = tokenize(self.fs, self.paths, self.part_size, columns)
+        read_avro_name = "read-avro-partition-" + token
+        dsk = {
+            (read_avro_name, i): (AvroDatasetEngine.read_partition, self.fs, piece, columns)
+            for i, piece in enumerate(pieces)
+        }
+        return new_dd_object(dsk, read_avro_name, meta.iloc[:0], [None] * (len(pieces) + 1))
+
+    def process_metadata(self, columns=None):
+
+        with open(self.paths[0], "rb") as fo:
+            header = ua.core.read_header(fo)
+
+            # Use first block for metadata
+            num_rows = header["blocks"][0]["nrows"]
+            file_byte_count = header["blocks"][0]["size"]
+            meta = cudf.io.read_avro(self.paths[0], skiprows=0, num_rows=num_rows)
+
+            # Convert the desired in-memory GPU size to the expected
+            # on-disk storage size (blocksize)
+            df_byte_count = meta.memory_usage(deep=True).sum()
+            self.blocksize = int(float(file_byte_count) / df_byte_count * self.part_size)
+
+        # Break apart files at the "Avro block" granularity
+        pieces = []
+        for path in self.paths:
+            file_size = self.fs.du(path)
+            if file_size > self.blocksize:
+                with open(path, "rb") as fo:
+                    header = ua.core.read_header(fo)
+                    ua.core.scan_blocks(fo, header, file_size)
+                    blocks = header["blocks"]
+
+                    (file_row_offset, part_row_count) = (0, 0)
+                    (file_block_offset, part_block_count) = (0, 0)
+                    (file_byte_offset, part_byte_count) = (blocks[0]["offset"], 0)
+
+                    for i, block in enumerate(blocks):
+                        part_row_count += block["nrows"]
+                        part_block_count += 1
+                        part_byte_count += block["size"]
+                        if part_byte_count >= self.blocksize:
+                            pieces.append(
+                                {
+                                    "path": path,
+                                    "rows": (file_row_offset, part_row_count),
+                                    "blocks": (file_block_offset, part_block_count),
+                                    "bytes": (file_byte_offset, part_byte_count),
+                                }
+                            )
+                            (file_row_offset, part_row_count) = (
+                                file_row_offset + part_row_count,
+                                0,
+                            )
+                            (file_block_offset, part_block_count) = (
+                                file_block_offset + part_block_count,
+                                0,
+                            )
+                            (file_byte_offset, part_byte_count) = (
+                                file_byte_offset + part_byte_count,
+                                0,
```

Comment:
I think this can be more cleanly expressed without tuple unpacking - something like 
```suggestion
                            file_row_offset += part_row_count
                            file_block_offset += part_block_count
                            file_byte_offset += part_byte_count
                            part_row_count = part_row_count = part_block_count = 0
```